src/openms/source/QC/Identifications.cpp

```diff
@@ -0,0 +1,125 @@
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Axel Walter $
+// $Authors: Axel Walter $
+// --------------------------------------------------------------------------
+
+
+#include <OpenMS/QC/Identifications.h>
+#include <OpenMS/FORMAT/IdXMLFile.h>
+#include <iostream>
+using namespace std;
+
+namespace OpenMS
+{ 
+
+  Identifications::Result Identifications::compute(const String& inputfile_id)
+  {
+    Identifications::Result result;
+    vector<ProteinIdentification> prot_ids;
+    vector<PeptideIdentification> pep_ids;
+    IdXMLFile().load(inputfile_id, prot_ids, pep_ids);
+    set<String> peptides;
+    set<String> proteins;
+    // PSMs and collect unique peptides in set
+    for (const auto& pep_id : pep_ids)
+    {
+      if (!pep_id.empty())
+      {
+        result.peptide_spectrum_matches += pep_id.getHits().size();
+        const auto& temp_hits = pep_id.getHits();
+        for (const auto& temp_hit : temp_hits)
+        {
+          peptides.insert(temp_hit.getSequence().toString());
+        }
+      }
+    }
+    // get sum of all peptide length for mean calculation
+    // get missed cleavages, check if Lys or Arg in the middle of sequence
+    UInt missed_cleavages = 0;
+    UInt peptide_length_sum = 0;
+    for (const auto& pep : peptides)
+    {
+      peptide_length_sum += pep.size();
+      for (const auto& aa : pep)
+      {
+        if (aa == 'K' || aa == 'R')
+        {
+          ++missed_cleavages;
+        }
+      }
+    }
+    result.peptide_length_mean = (double)peptide_length_sum / peptides.size();
+    result.missed_cleavages_mean = (double)missed_cleavages / peptides.size();
+    // collect unique proteins in sets and scores mean
+    double protein_hit_scores_sum = 0;
+    UInt protein_hit_count = 0;
+    for (const auto& prot_id : prot_ids)
+    {
+      const auto& temp_hits = prot_id.getHits();
```

Comment:
this is tricky, because if the pepIDs contain multiple peptidehits, you'd probably want to exclusively retain the proteins which are references in the best peptideHit and ignore the rest.
Currently this is a vast overestimation.